runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -6263,26 +6262,37 @@ done:;
 		void *resolveResult = NULL;
 
 		if (J9_UNEXPECTED(!VM_VMHelpers::staticFieldRefIsResolved(flagsAndClass, valueOffset, true))) {
-			/* Unresolved */
-			J9Method *method = _literals;
-			buildGenericSpecialStackFrame(REGISTER_ARGS, 0);
-			updateVMStruct(REGISTER_ARGS);
-			resolveResult = resolveStaticFieldRef(_currentThread, method, ramConstantPool, index, J9_RESOLVE_FLAG_RUNTIME_RESOLVE | J9_RESOLVE_FLAG_FIELD_SETTER | J9_RESOLVE_FLAG_CHECK_CLINIT, NULL);
-			VMStructHasBeenUpdated(REGISTER_ARGS);
-			restoreGenericSpecialStackFrame(REGISTER_ARGS);
-			if (immediateAsyncPending()) {
-				rc = GOTO_ASYNC_CHECK;
-				goto done;
-			} else if (VM_VMHelpers::exceptionPending(_currentThread)) {
-				rc = GOTO_THROW_CURRENT_EXCEPTION;
-				goto done;
-			}
-			if ((void*)-1 != resolveResult) {
-				goto retry;
+			/* The above test will fail if the ref has been resolved for put and the target
+			 * field is final. In this case, ensure that the current method is a constructor
+			 * if the class is enforcing the final field setting rules.
+			 *
+			 * If the check fails, run the resolve again to throw the exception.
+			 */
+			if (!J9_ARE_ALL_BITS_SET(flagsAndClass, (UDATA)(J9StaticFieldRefPutResolved | J9StaticFieldRefFinal) << (8 * sizeof(UDATA) - J9_REQUIRED_CLASS_SHIFT))
+				|| (VM_VMHelpers::romClassChecksFinalStores(ramConstantPool->ramClass->romClass)
+					&& !VM_VMHelpers::romMethodIsConstructor(J9_ROM_METHOD_FROM_RAM_METHOD(_literals), false)
+				   )
```

Comment:
Not a fan of this formatting but also not a fan of any alternative I've been able to come up with

Reply:
Agreed - it looks better in Eclipse.